### PR TITLE
[CI] Mac OS app

### PR DESCRIPTION
**Pull request by pazos**
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6955)

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]
Sure. :-)

**Comment by pazos** [comment]
wooops. `$VERSION` doesn't work, so `git-rev` is empty. No idea how to fix that :/

**Comment by NiLuJe** [comment]
Which part doesn't work, exactly?

**Comment by pazos** [comment]
@NiLuJe https://github.com/koreader/koreader/blob/master/kodev#L11

**Comment by NiLuJe** [comment]
Is this a shallow clone, perhaps?

**Comment by Frenzie** [comment]
What's the error message? But yeah, best to do at least 100 or 200 deep if
you want that to work.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        os: [macos-latest]
+
+    steps:
+      - name: XCode version
+        run: xcode-select -p
+
+      - name: Check out Git repository
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
```
Comment (Frenzie) [inline .github/workflows/build.yml:20]: You don't need 0, something like 100 (and tags if necessary) will likely do, probably better 200 to play it safe. Our total number of commits is close to 10000, but by limiting it to what's likely to happen within a couple of months you still get most of the speed advantages of a shallow clone. (I.e., it'll take ~1 second as opposed to ~4. Maybe not worth the optimization when the build itself takes over 10 minutes, but just fyi anyway. :-P)

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+      - name: Cache for ccache
+        uses: actions/cache@v1
+        env:
+          cache-name: cache-ccache
+        with:
+          path: ~/.ccache # ccache cache files are stored in `~/.ccache` on Linux/macOS
+          key: ${{ runner.os }}-build-${{ env.cache-name }}
```
Comment (Frenzie) [inline .github/workflows/build.yml:28]: If you want to run this on every PR, you should cache base since that's only updated occasionally. See my CircleCI config.

https://github.com/koreader/koreader/blob/c20ad8f5e9385dd18138a306b874c5b20302f299/.circleci/config.yml#L20-L26

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+          restore-keys: |
+            ${{ runner.os }}-build-${{ env.cache-name }}-
+            ${{ runner.os }}-build-
+            ${{ runner.os }}-
```
Comment (Frenzie) [inline .github/workflows/build.yml:32]: Incidentally, if you copied this from base it's probably not quite right yet. But with the macOS build constantly breaking itself I haven't really been able to look into it.

File: .github/workflows/build.yml (hunk 2 above)
Comment (pazos) [reply .github/workflows/build.yml:28]: I think running this on tagged releases is enough for https://github.com/koreader/koreader/issues/6875. No idea how to do that, how artifacts work and, generally speaking, the work involved on each release. So I would be happy if you could take care of it and remain ignorant :)

File: .github/workflows/build.yml (hunk 3 above)
Comment (pazos) [reply .github/workflows/build.yml:32]: Yup, totally ripped from base. Remember I have no idea of what I'm doing here :p

File: .github/workflows/build.yml (hunk 1 above)
Comment (pazos) [reply .github/workflows/build.yml:20]: Happy to change it :+1:

**Comment by pazos** [comment]
@Frenzie: I don't speak much git, github, automatisms or build systems.  It is all a bit magic to me :smiley: 

I was trying to address your comment in https://github.com/koreader/koreader/issues/6875#issuecomment-727204973.

File: .github/workflows/build.yml (hunk 2 above)
Comment (Frenzie) [reply .github/workflows/build.yml:28]: In that case you can add a condition like https://github.com/hellerbarde/stapler/commit/48d2dc7e7a683ca3c2b659c1762aa67661f7ad79 

For tags/releases only we don't need to cache base.

It can be slightly annoying to test, but it's not too much of a bother on your own fork.

Anyhow, I can also do that in a follow-up PR some other time.

File: .github/workflows/build.yml (hunk 3 above)
Comment (Frenzie) [reply .github/workflows/build.yml:32]: It's the same general idea as in CircleCI. The cache is a never-changing thing (maybe it's deleted if it's not used for a few months?), so to generate an up to date cache it needs to be sufficiently unique.

https://github.com/koreader/koreader/blob/c20ad8f5e9385dd18138a306b874c5b20302f299/.circleci/config.yml#L19

The problem is that I don't think `${{ runner.os }}-build-${{ env.cache-name }}-` has actually changed since I added the macOS build. So the build time is slowly increasing as the cache becomes increasingly unrelated to the repo.

**Comment by Frenzie** [comment]
> I was trying to address your comment in #6875 (comment).

Oh sure, if it works Good Enough™ for the moment I'm fine with merging it as also indicated by the approval I left.

**Comment by pazos** [comment]
@Frenzie


```
      - name: Check out Git repository
        uses: actions/checkout@v2
        with:
          fetch-depth: 200
```

doesn't work. As you can see in the artifacts we lost $VERSION again.

**Comment by Frenzie** [comment]
By "again" you mean it worked at some point? If you don't have tags... ;-)

**Comment by pazos** [comment]
> By "again" you mean it worked at some point?

It worked with `fetch-depth: 0`.
And by worked I mean the osx-artifacts.zip contained a valid version and a working program inside :100:

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    steps:
+      - name: XCode version
+        run: xcode-select -p
+
+      - name: Check out Git repository
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 200
+
```
Comment (Frenzie) [inline .github/workflows/build.yml:21]: try adding:
```
run: git fetch --tags
```

Comment (Frenzie) [reply .github/workflows/build.yml:21]: Or just go with 0. It's just 3 seconds. ;-)

Comment (NiLuJe) [reply .github/workflows/build.yml:21]: (... and not our bandwidth. Mwhahaha.)

**Comment by pazos** [comment]
@JasonInOttawa: could you test https://github.com/koreader/koreader/suites/1645100449/artifacts/30493747 on your pristine mac?. It seems to work fine on mine.

**Comment by JasonInOttawa** [comment]
A quick download and run worked fine. I got an osx-artifacts.zip file, which contained a .7z file, which contained the application.

I did notice that it's not reading my settings - are they still stored in .config/koreader?

Thanks!

**Comment by JasonInOttawa** [comment]
@pazos Found it in Application Support - very nice!

**Comment by JasonInOttawa** [comment]
The test app has been happily running for a few hours now. I do have this computer set up for builds - could that make any difference?

**Comment by pazos** [comment]
> The test app has been happily running for a few hours now. I do have this computer set up for builds - could that make any difference?

I tested the app in a pristine VM (without brew, xcode or commandline tools). It works fine there. The reason to ask for your feedback is because the VM lacks some security enhancements available on apple hardware (namely System Integrity Protection, aka SIP).

The only reason that prevents us from releasing a wide mac app (with a minimum target of 10.09) is that we're relying on some brew libraries and they're built with newer deployment targets.

So, yeah. It should work fine in Big Sur and Catalina and might work on Mojave too.

**Comment by JasonInOttawa** [comment]
I'm running Big Sur. I did have to right-click - Open - Yes I'm sure as the app isn't notarized. No issues other than that.

Thanks!

**Comment by pazos** [comment]
> I did have to right-click - Open - Yes I'm sure as the app isn't notarized. No issues other than that.

That's the gatekeeper, not SIP. Based on your report SIP issues were fixed with the osx-loader in https://github.com/koreader/koreader/pull/6448#issuecomment-665345741

About the gatekeeper issue. Notarization is a joke for a few reasons:

- Notarized apps still require the same workaround as non notarized apps in default osx setup (gatekeeper enforced)
- Notarized apps require a [hardened runtime](https://developer.apple.com/documentation/security/hardened_runtime). Easy to implement in XCode projects but a PITA elsewhere. Also a hardened runtime disables JIT so we need to add an entitlement to disable that specific protection, maybe others too.
- Notarized apps require a **paid developer account**. Lol

tl;dr: notarization is a marketing thing, unrelated to security, which severly restricts free development for a mac. SIP is a security feature, unrelated to marketing, that doesn't harm free development and it is a cool thing to have it enabled.

**Comment by JasonInOttawa** [comment]
Got it.

It certainly would seem to be a good idea to waive the developer account fee for certain developers on projects that have been around for a while. If I had more pull at Apple, I'd suggest it :)

Looking forward to downloading the app from the Releases page one day!
